src/Debugger/Impl/rtvs/R/grid.R

```diff
@@ -72,3 +72,79 @@ grid.dput <- function(obj) {
     json;
 }
 
+getEnvRepr <- function(env, startEnvLevel) {
+  envRepr <- NULL;
+
+  if (startEnvLevel > 4L) {    # if within call or unknown, try to get call frame
+    nframe <- sys.nframe();
+    if (nframe > 0) {
+      for (i in 1:nframe) {
+        frame <- sys.frame(i)
+        if (identical(env, frame)) {
+          envRepr <- list(name = deparse(sys.call(i)[[1]]), level = 5L, frameIndex = i);
+          break;
+        }
+      }
+    }
+  }
+
+  if (is.null(envRepr)) {
+    if (identical(env, baseenv())) {
+      envRepr <- list(name = 'package:base', level = 2L);
+    } else if (identical(env, globalenv())) {
+      envRepr <- list(name = '.GlobalEnv', level = 4L);
+    } else if (identical(env, emptyenv())) {
+      envRepr <- list(name = 'EmptyEnv', level = 1L);
+    } else {
+      envRepr <- list(name = environmentName(env), level = 3L);
+    }
+  }
+  envRepr;
+}
+
+# return list of environment statck walking from the given environment up.
+# R environment level: empty(1) > base(2) > packages(3) > global(4) > calls(5) > unknown(10)
+getEnvironments <- function(env) {
+  if (missing(env)) {
+    curEnv <- sys.frame(-1);
+  } else {
+    curEnv <- env;
+  }
+  envs <- list();
+
+  prevEnvLevel <- 10L;
```

Comment:
Okay, so these are "levels" from the comment above? But then it is wrong. The order that you present is the one that `loadLibrary` / `library` will create, but there's also `data`, and arbitrary environments can be inserted into search path, so this hierarchy of levels does not hold in practice. Besides, there's no reason to distinguish between them at this point. The only worthwhile distinction is search path vs call stack.